### PR TITLE
Simplify system.DiskUsage()

### PR DESCRIPTION
**Pull request by ndenev**
* skip over filesystems in the ignore list before collecting stats for them
 * disk.Usage() can fail even if os.Stat() succeeds on the mountpoint (remote filesytems, autofs, etc). In this case skip over the affected filesystem and not just return error breaking stats for all other filesystems.

### Required for all PRs:

- [ ] CHANGELOG.md updated (we recommend not updating this until the PR has been approved by a maintainer)
- [x] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [ ] README.md updated (if adding a new plugin)

### CODE REVIEW

File: plugins/inputs/system/ps.go [hunk 1]
```diff
@@ -73,35 +73,27 @@ func (s *systemPS) DiskUsage(
...
-			du.Path = p.Mountpoint
-
-			// If the mount point is a member of the exclude set,
-			// don't gather info on it.
-			_, ok := fstypeExcludeSet[p.Fstype]
-			if ok {
-				continue
+			if du, err := disk.Usage(mountpoint); err == nil {
+				du.Path = p.Mountpoint
+				du.Fstype = p.Fstype
```
Comment (ndenev) [inline plugins/inputs/system/ps.go]: I'm not sure also why these are needed. On my Machine (Mac) they are always the same so this looks pointless.

Comment (danielnelson) [inline plugins/inputs/system/ps.go]: `Path` and `Fstype` are the same?

Comment (ndenev) [inline plugins/inputs/system/ps.go]: I meant I see Path has the same value as Mountpoint.
Also Fstype field in "p" has the same value as in "du".
p is of disk.PartitionStat type and du is disk.UsageStat : https://godoc.org/github.com/shirou/gopsutil/disk#PartitionStat

That's why I don't see why are they assigned.
Unless I'm missing something I can update the PR to remove those two assignments if they are indeed redundant.

Comment (danielnelson) [inline plugins/inputs/system/ps.go]: I think it may be something related to docker support.  We need to be careful with changes here as proper unit tests do not exist.

File: plugins/inputs/system/ps.go [hunk 2]
```diff
@@ -73,35 +73,27 @@ func (s *systemPS) DiskUsage(
...
 
-	for i := range parts {
-
-		p := parts[i]
+	for _, p := range parts {
+		// If the mount point is not a member of the filter set,
+		// don't gather info on it.
+		if _, ok := mountPointFilterSet[p.Mountpoint]; !ok {
+			continue
+		}
```
Comment (phemmer) [inline plugins/inputs/system/ps.go]: If `mountPointFilter` is empty (which is the config default, and means to gather everything), this is going to not gather anything.

File: plugins/inputs/system/ps.go [hunk 3]
```diff
@@ -73,35 +73,27 @@ func (s *systemPS) DiskUsage(
...
-			// don't gather info on it.
-			_, ok := mountPointFilterSet[p.Mountpoint]
-			if !ok {
-				continue
-			}
+		// If the mount point is a member of the exclude set,
+		// don't gather info on it.
+		if _, ok := fstypeExcludeSet[p.Fstype]; ok {
+			continue
 		}
```
Comment (phemmer) [inline plugins/inputs/system/ps.go]: Similar to above. If `fstypeExclude` is empty (again default, collect everything), this is not going to gather anything.

File: plugins/inputs/system/ps.go [hunk 4]
```diff
@@ -73,35 +73,27 @@ func (s *systemPS) DiskUsage(
...
-				return nil, nil, err
-			}
-			du.Path = p.Mountpoint
-
-			// If the mount point is a member of the exclude set,
-			// don't gather info on it.
-			_, ok := fstypeExcludeSet[p.Fstype]
-			if ok {
-				continue
+			if du, err := disk.Usage(mountpoint); err == nil {
```
Comment (phemmer) [inline plugins/inputs/system/ps.go]: If we're trying to be idiomatic, idiomatic go is to check for `err != nil` and if so `return`/`continue`. This is so that you don't have to do so much nesting.

File: plugins/inputs/system/ps.go (hunk 2 above)
Comment (phemmer) [inline plugins/inputs/system/ps.go]: Ack, accidentally deleted the wrong comment, so repeating what I deleted:

If `mountPointFilter` is empty (which is the default and means to collect everything), this is going to collect nothing.

File: plugins/inputs/system/ps.go (hunk 1 above)
Comment (phemmer) [inline plugins/inputs/system/ps.go]: It's because of things like this:

```
rootfs / rootfs rw 0 0
/dev/mapper/centos-root / xfs rw,relatime,attr2,inode64,noquota 0 0
```

The `for ... range parts {` is going to iterate over both these. However when the `disk.Usage()` call runs for the first one, it's going to get the second one because it's mounted on top. So `p.Fstype == "rootfs"` but `du.Fstype == "xfs"`.

Comment (ndenev) [inline plugins/inputs/system/ps.go]: Ok, I see. I'm leaving these alone then.

File: plugins/inputs/system/ps.go (hunk 4 above)
Comment (ndenev) [inline plugins/inputs/system/ps.go]: Ok.

File: plugins/inputs/system/ps.go (hunk 2 above)
Comment (ndenev) [inline plugins/inputs/system/ps.go]: Ah, that's right. I think I was testing locally with non empty config. I'll fix that and update the PR.

**Comment by danielnelson** [comment]
Can you add unittests?  In case we introduce a bug with this refactor I want to have tests in place so we can move towards locking the behavior down.

**Comment by ndenev** [comment]
There are tests for this in disk_test.go.
I'll see if I can add some more.

**Comment by danielnelson** [comment]
I think it is using a testing implementation.

**Comment by ndenev** [comment]
You are right, it's a mock implementation of `DiskUsage{}`.
So, I'm now trying to see what's the best way to unit test this as I'm still learning Go, and I wan't able to make it work using `mock_PS` as there is already mock implementation of `DiskUsage` and I need just to mock a few calls to `os` and `gopsutil` from within the real `DiskUsage` func/method.

I came up with this [https://github.com/influxdata/telegraf/compare/master...ndenev:diskstat-tests](https://github.com/influxdata/telegraf/compare/master...ndenev:diskstat-tests)
Basically moving all the `gopsutil` and `os` calls that `DiskUsage{}` makes as package level function pointers that can be mocked from inside the test. Hopefully the compiler is smart enough to optimize these out in the binary.

I'm also open to suggestions of better approach :)

P.S.: The new test data is just copy of the existing higher level test, but I plan to import some test data from more modern distro having all of the cgroup, etc filesystems.

**Comment by danielnelson** [comment]
Instead of function pointers I would make an interface.  It's possible that the PS interface can be modified and used, or maybe it needs to be a new interface.
